tests/msc2403_test.go

```diff
@@ -0,0 +1,281 @@
+// +build msc2403
+
+package tests
+
+import (
+	"encoding/json"
+	"net/url"
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/tidwall/gjson"
+)
+
+// A reason to include in the request body when testing knock reason parameters
+var testKnockReason string = "Let me in... LET ME IN!!!"
+
+// TestKnocking tests sending knock membership events and transitioning from knock to other membership states.
+// Knocking is currently an experimental feature and not in the matrix spec.
+// This function tests knocking on local and remote room.
+func TestKnocking(t *testing.T) {
+	deployment := Deploy(t, "test_knocking", b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	// Create a client for one local user
+	aliceUserID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", aliceUserID)
+
+	// Create a client for another local user
+	bobUserID := "@bob:hs1"
+	bob := deployment.Client(t, "hs1", bobUserID)
+
+	// Create a client for a remote user
+	charlieUserID := "@charlie:hs2"
+	charlie := deployment.Client(t, "hs2", charlieUserID)
+
+	// Create a room for alice and bob to test knocking with
+	roomIDOne := alice.CreateRoom(t, struct {
+		Preset      string `json:"preset"`
+		RoomVersion string `json:"room_version"`
+	}{
+		"private_chat",      // Set to private in order to get an invite-only room
+		"xyz.amorgan.knock", // Room version required for knocking. TODO: Remove when knocking is in a stable room version
```

Comment:
What do we do when this is standardized? Do we just update these? Do we somehow let a server specify that they support the MSC vs. the standardized version?

Reply:
Once the MSC is merged into a released version of the spec, Complement should be updated to support that release. Part of that update will be stabilizing knocks.

The necessary steps for stabilizing will be:

* Updating instances of `xyz.amorgan.knock` to `knock`.
* Update endpoints from `/_matrix/client/unstable/xyz.amorgan.knock/...` to `/_matrix/client/rX/...`.
* Removing the `// +build ...` line at the top so the tests are run by default (they're currently opt in).
* Updating the name of the test file to `knock_test.go` or somesuch.
* Removing the conditional on rescinding knocks over federation. Synapse will need to support that if the spec says it's possible.

It might be a good idea to put the above list into the top of the file actually.

Reply:
I guess my broader question is -- does complement have a way for homeservers to say what version of the spec they support? Not expecting it here necessarily, just bringing it up for conversation!

And if it does, does it make sense for a homeserver to say it supports 0.6 + msc2403 even after 0.7 is released?

Reply:
@Kegsay I'd be interested in your answer here in case it differs from mine.

I don't think it does, and in doing so I think Complement would need to support multiple spec versions, which I don't think it's planning to. Mostly because new spec versions often require modifying existing tests instead of making a separate set of new ones.

Selecting a range of tests is done via build tags, thus I think this choice would be made in your individual CI script rather than in a config file in the server/complement (at least currently!).

Reply:
Alright. I don't think we need to solve this here, was just a thought I had that!